### PR TITLE
plg/process: Cache hash file

### PR DESCRIPTION
**Pull request by marselester**
An attempt to tackle https://github.com/parca-dev/parca-agent/issues/1100.

```
# new
BenchmarkPIDMappingFileCache-2   	  135170	      8710 ns/op	    3468 B/op	      32 allocs/op
# old
BenchmarkPIDMappingFileCache-2   	   16863	     76147 ns/op	   36268 B/op	      33 allocs/op
```

### CODE REVIEW

**Comment by javierhonduco** [comment]
Thanks so much for working on this PR! The improvements are significant! That being said, I am wary of adding caching layers before exhausting all the other options. Caching introduces complexity, making it harder to debug issues, increases memory usage and might result in reducing the visibility of the real costs.

Before we move forward with this PR I would suggest exploring and benchmarking other hashing algorithms, but let's see what other Parca Agent maintainers have to see on this!

**Comment by brancz** [comment]
After adapting Parca Agent to follow the new signed URL protocol (#1104), it seems that this is no longer an issue. A version with that patch is already running on the demo instance, and it seems that that made it go down to ~1%. Judging by what I'm seeing on demo I think we have much bigger fish to fry right now (allocations in EnsureUploaded and profiler loop stacks look like they should give us a major wins).

**Comment by brancz** [comment]
I think what @javierhonduco mentioned does make sense though, it also looks like we checked in a highwayhash based hashing into this repo but didn't actually end up using that (see `/pkg/hash` in parca agent). Maybe we can benchmark this again against Parca's `xxhash` based hashing. I believe we benchmarked this before and added highwayhash because of that, but I don't recall anymore.

**Comment by marselester** [comment]
No worries. I thought that the issue https://github.com/parca-dev/parca-agent/issues/1100 is still relevant since it was created a week ago.

> it also looks like we checked in a highwayhash based hashing into this repo but didn't actually end up using that

It's been used in the following files.

```
./pkg/ksym/ksym.go:	"github.com/parca-dev/parca-agent/pkg/hash"
./pkg/debuginfo/manager.go:	"github.com/parca-dev/parca/pkg/hash"
./pkg/perf/perf.go:	"github.com/parca-dev/parca-agent/pkg/hash"
./pkg/process/maps.go:	"github.com/parca-dev/parca-agent/pkg/hash"
```

**Comment by kakkoyun** [comment]
@marselester Related https://github.com/parca-dev/parca-agent/issues/1139#issuecomment-1357696135

File: cmd/parca-agent/main.go [hunk 1]
```diff
@@ -340,6 +340,10 @@ func run(logger log.Logger, reg *prometheus.Registry, flags flags) error {
 		flags.ProfilingDuration, // Cache durations are calculated from profiling duration.
 	)
 
+	c, err := process.NewMappingFileCache(logger)
```
Comment (kakkoyun) [inline cmd/parca-agent/main.go:343]:
```suggestion
	mappingFileCache, err := process.NewMappingFileCache(logger)
```

**Review by kakkoyun — COMMENTED** [review]
looking good 👍

File: pkg/process/maps.go [hunk 2]
```diff
@@ -72,24 +91,51 @@ func (c *mappingFileCache) MappingForPID(pid int) ([]*profile.Mapping, error) {
...
+	// This is even though files are extremely unlikely to change.
+	// We should only rehash if the mtime of the file changes.
+	var (
+		t int64
+		h uint64
+		r io.Reader
+	)
+	stat, err := f.Stat()
+	if err == nil && stat != nil {
+		t = stat.ModTime().UnixNano()
```
Comment (brancz) [inline pkg/process/maps.go:115]: do we have some docs or proof that `procfs` "files" (in particular mappings here of course) actually behave as expected in regards to mtime?

Comment (brancz) [reply pkg/process/maps.go:115]: this StackOverflow post suggests we cannot trust it: https://stackoverflow.com/a/30885400

Comment (marselester) [reply pkg/process/maps.go:115]: You're right! All the processes on my Linux machine but pid 1 show atime/mtime/ctime matching a time of the first `open` or `stat`.

```bash
$ stat /proc/2/maps
  File: /proc/2/maps
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 16h/22d	Inode: 397849      Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-12-21 16:34:11.553962763 +0000
Modify: 2022-12-21 16:34:11.553962763 +0000
Change: 2022-12-21 16:34:11.553962763 +0000
 Birth: -

$ stat /proc/1/maps
  File: /proc/1/maps
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 16h/22d	Inode: 25064       Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-11-29 21:26:10.799570796 +0000
Modify: 2022-11-29 21:26:10.799570796 +0000
Change: 2022-11-29 21:26:10.799570796 +0000
 Birth: -

$ sudo go run ./t/
2022-11-29 21:26:10.799570796 +0000 UTC
```

<details>

```go
package main

import (
	"fmt"
	"os"
)

func main() {
	f, err := os.Open("/proc/1/maps")
	if err != nil {
		panic(err)
	}
	s, err := f.Stat()
	if err != nil {
		panic(err)
	}
	fmt.Println(s.ModTime())
}
```

</details>

**Comment by zdyj3170101136** [comment]
> 不用担心。我认为问题#1100仍然相关，因为它是在一周前创建的。
> 
> > 看起来我们还检查了一个基于 highwayhash 的哈希到这个 repo 但实际上并没有最终使用它
> 
> 它已在以下文件中使用。
> 
> ```
> ./pkg/ksym/ksym.go:	"github.com/parca-dev/parca-agent/pkg/hash"
> ./pkg/perf/perf.go:	"github.com/parca-dev/parca-agent/pkg/hash"
> ./pkg/process/maps.go:	"github.com/parca-dev/parca-agent/pkg/hash"
> ```

could add it to 
```
debuginfo/manager.go function exists.
debuginfo/upload.go
```

File: pkg/process/maps.go (hunk 2 above)
Comment (zdyj3170101136) [reply pkg/process/maps.go:115]: > 你是对的！我的 Linux 机器上除 pid 1 之外的所有进程都显示 atime/mtime/ctime 匹配第一个时间`open`或`stat`.
> 
> ```shell
> $ stat /proc/2/maps
>   File: /proc/2/maps
>   Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
> Device: 16h/22d	Inode: 397849      Links: 1
> Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
> Access: 2022-12-21 16:34:11.553962763 +0000
> Modify: 2022-12-21 16:34:11.553962763 +0000
> Change: 2022-12-21 16:34:11.553962763 +0000
>  Birth: -
> 
> $ stat /proc/1/maps
>   File: /proc/1/maps
>   Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
> Device: 16h/22d	Inode: 25064       Links: 1
> Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
> Access: 2022-11-29 21:26:10.799570796 +0000
> Modify: 2022-11-29 21:26:10.799570796 +0000
> Change: 2022-11-29 21:26:10.799570796 +0000
>  Birth: -
> 
> $ sudo go run ./t/
> 2022-11-29 21:26:10.799570796 +0000 UTC
> ```
> 
> ```go
> package main
> 
> import (
> 	"fmt"
> 	"os"
> )
> 
> func main() {
> 	f, err := os.Open("/proc/1/maps")
> 	if err != nil {
> 		panic(err)
> 	}
> 	s, err := f.Stat()
> 	if err != nil {
> 		panic(err)
> 	}
> 	fmt.Println(s.ModTime())
> }
> ```

i think the /proc/pid/maps file's mod time is expected to equal with access time.

and it is not allowed to modify:
```
sudo chmod 0755 /proc/19177/maps
chmod: changing permissions of ‘/proc/19177/maps’: Operation not permitted
```

i have a log file, the stat output meets i expect.
```
$ stat otel-collector.log
  File: ‘otel-collector.log’
  Size: 9145273667      Blocks: 17861872   IO Block: 4096   regular file
Device: fd11h/64785d    Inode: 1835016     Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-12-20 17:24:30.243937079 +0800
Modify: 2022-12-22 12:10:03.450355314 +0800
Change: 2022-12-22 12:10:03.450355314 +0800
 Birth: -
$ stat otel-collector.log
  File: ‘otel-collector.log’
  Size: 9145282651      Blocks: 17861896   IO Block: 4096   regular file
Device: fd11h/64785d    Inode: 1835016     Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-12-20 17:24:30.243937079 +0800
Modify: 2022-12-22 12:10:05.482427849 +0800
Change: 2022-12-22 12:10:05.482427849 +0800
 Birth: -
$ pwd
/proc/19177/root/home/data/logs/takumi
```

**Pull request closed without merging**